src/coreir-c/simulator-c.cpp

```diff
@@ -0,0 +1,116 @@
+#include "coreir-c/coreir-simulator.h"
+#include "coreir.h"
+#include "common-c.hpp"
+#include "../simulator/interpret.hpp"
+
+using namespace std;
+namespace CoreIR {
+
+static vector<string> MakeSimPath(char** cpath, int len) {
+  vector<string> path;
+  for (int i = 0; i < len; i++) {
+    path.emplace_back(cpath[i]);
+  }
+
+  return path;
+}
+
+extern "C" {
+
+  bool CORESimValueGetBit(CORESimValue* cval, int bit) {
+    SimValue* val = rcast<SimValue*>(cval);
+    if (val->getType() == SIM_VALUE_BV) {
+      SimBitVector* bv = static_cast<SimBitVector*>(val);
+      return bv->getBits().get(bit);
+    }
+    else if (val->getType() == SIM_VALUE_CLK) {
+      ClockValue* cv = static_cast<ClockValue*>(val);
+      return cv->value();
+    }
+    return false;
+  }
+
+  int CORESimValueGetLength(CORESimValue* cval) {
+    SimValue* val = rcast<SimValue*>(cval);
+    if (val->getType() == SIM_VALUE_BV) {
+      SimBitVector* bv = static_cast<SimBitVector*>(val);
+      return bv->getBits().bitLength();
+    } else if (val->getType() == SIM_VALUE_CLK) {
+      return 1;
+    }
+    return 0;
+  }
+
+  CORESimulatorState* CORENewSimulatorState(COREModule* module) {
+    Module* top = rcast<Module*>(module);
+    SimulatorState* state = new SimulatorState(top);
+
+    return rcast<CORESimulatorState*>(state);
+  }
+
+  void COREDeleteSimulatorState(CORESimulatorState* state) {
+    delete rcast<SimulatorState*>(state);
+  }
+
+  CORESimValue* CORESimGetValue(CORESimulatorState* cstate, char** cpath, int path_len) {
+    SimulatorState* state = rcast<SimulatorState*>(cstate);
+    vector<string> path = MakeSimPath(cpath, path_len);
+
+    return rcast<CORESimValue*>(state->getValue(path));
+  }
+
+  void CORESimSetMainClock(CORESimulatorState* cstate, char** cpath, int path_len) {
+    SimulatorState*state = rcast<SimulatorState*>(cstate);
+    vector<string> path = MakeSimPath(cpath, path_len);
+    state->setMainClock(path);
+  }
+
+  void CORESimSetClock(CORESimulatorState* cstate, char** cpath, int path_len, bool lastval, bool curval) {
+    SimulatorState* state = rcast<SimulatorState*>(cstate);
+    vector<string> path = MakeSimPath(cpath, path_len);
+
+    state->setClock(path, lastval, curval);
+  }
+
+  void CORESimSetValue(CORESimulatorState* cstate, char** cpath, int path_len, bool* new_val, int val_len) {
+    SimulatorState* state = rcast<SimulatorState*>(cstate);
+    vector<string> path = MakeSimPath(cpath, path_len);
+    BitVec bv(val_len);
+    for (int i = 0; i < val_len; i++) {
+      bv.set(i, new_val[i]);
+    }
+    state->setValue(path, bv);
+  }
+
+  void CORESimStepMainClock(CORESimulatorState* cstate) {
+    SimulatorState* state = rcast<SimulatorState*>(cstate);
+    state->stepMainClock();
+  }
```

Comment:
It seems like wrapping method invocations on classes with no arguments or return values could be wrapped up in a simple script to generate it. We could create a simple script to manage these patterns and probably use it to generate the ctypes declaration for `pycoreir` as well. Maybe this is what SWIG does? Now that I'm thinking about it, I don't think it would be that hard to write a script to generate an implementation of the data-structure transfers from C++ to C to Python (e.g. methods that pass an UnorderedMap to key, value arrays in C to a python dictionary).

```
{{class}}.{{method}}
```

creates
```
void CORE{{class}}{{method}}(CORE{{class}}* arg) {
  rcast<{{class}}*>(arg)->{{method}};
}
```